### DISCUSSION TITLE
Atomic JSON updates with SQLAlchemy 1.4

### DISCUSSION THREAD

**Discussion by Amwam** [sqlalchemy/sqlalchemy#7806] [Usage Questions] [answered]
This follows on from a google group discussion https://groups.google.com/g/sqlalchemy/c/hjjIyEC8KHQ
We're attempting to solve the same problem in this discussion, and have adopted the solution described in there, in order to perform atomic JSONB updates.

The core of the code is essentially as recommended:

```python
DBSession.query(Map).filter(Map.id == id_).update(
    {"screenshots": func.jsonb_set(Map.screenshots, '{size}', '"filename.jpg"')}, 
    synchronize_session='fetch'
)
```
However we've adapted to make the code work with inheritance, and for columns that have synonyms.

I've publish a minimal example here with tests https://github.com/Amwam/sqlalchemy-jsonb-set-support
- Core function: https://github.com/Amwam/sqlalchemy-jsonb-set-support/blob/main/atomic_jsonb_set.py
- Tests: https://github.com/Amwam/sqlalchemy-jsonb-set-support/blob/main/test_atomic_jsonb.py

The tests all pass with 1.3, but fail in 1.4.

The error we get back is:
```
sqlalchemy.exc.ArgumentError: subject table for an INSERT, UPDATE or DELETE expected, 
    got Column('json_column', JSONB(astext_type=Text()), table=<parent_table>, nullable=False, default=ColumnDefault({})).
```

This does make sense as we're passing the column to `DBSession.query()`. 
Changing this to `context.__class__`/`mapper` (i.e. the table) the tests fail when trying to update the Child class. The emitted SQL targets the child table directly, despite the column being located on the Parent.

I'm unsure if this is a bug in SQLAlchemy 1.4, or if we're not creating this query correctly.

Thanks in advance for your advice!

Edit: Moved all the code into a single file
```python

import json

from sqlalchemy import func, inspect
from sqlalchemy import create_engine, Column, Integer, Unicode, ForeignKey
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import sessionmaker, scoped_session, synonym
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy_utils import database_exists, create_database

# ==============================
#  SETUP
# ==============================

engine = create_engine(
    "postgresql+psycopg2:///sql_test:sql_test@localhost:5432/sql_test", echo=True
)
if not database_exists(engine.url):
    create_database(engine.url)
DeclarativeBase = declarative_base()

maker = sessionmaker(autoflush=True, autocommit=False)
DBSession = scoped_session(maker)
DBSession.configure(bind=engine)


class Parent(DeclarativeBase):
    __tablename__ = "parent_table"

    id = Column(Integer, autoincrement=True, primary_key=True)
    json_column = Column(JSONB, nullable=False, default={})
    aliased_json_column = synonym("json_column")
    type = Column(Unicode, nullable=False)

    __mapper_args__ = {
        "polymorphic_on": type,
        "polymorphic_identity": "parent",
    }


class Child(Parent):
    __tablename__ = "child_table"

    child_id = Column(Integer, ForeignKey("parent_table.id"), primary_key=True)
    __mapper_args__ = {"polymorphic_identity": "child"}


DeclarativeBase.metadata.create_all(engine


# ==============================
# CORE
# ==============================

def atomic_jsonb_set(
        context: DeclarativeBase,
        dynamic_column_name: str,
        column_id,
        value,
):
    mapper = inspect(context.__class__)
    dynamic_column = getattr(context.__class__, dynamic_column_name)
    query = DBSession.query(dynamic_column)


    try:
        # needed for inheritance (persist_selectable is a Join)
        query = query.filter(mapper.persist_selectable.onclause)
    except AttributeError:
        # not a child class
        pass

    # filter by primary key
    id_key = [col for col in mapper.primary_key]
    for col in id_key:
        query = query.filter(col == getattr(context, col.name))

    query.update(
        {
            dynamic_column: func.jsonb_set(
                dynamic_column,
                f"{{{column_id}}}",
                json.dumps(value),
            )
        },
        synchronize_session="fetch",
    )

# ==============================
#  TEST
# ==============================

def _create_obj(cls):
    model = cls()
    DBSession.add(model)
    DBSession.flush()
    return model

def test_can_update_parent_json_column():
    parent = _create_obj(Parent)
    atomic_jsonb_set(
        parent, dynamic_column_name="json_column", column_id="foo", value="bar"
    )
    DBSession.refresh(parent)
    assert parent.aliased_json_column["foo"] == "bar"
    assert parent.json_column["foo"] == "bar"


def test_can_update_child_json_column():
    child = _create_obj(Child)
    atomic_jsonb_set(
        child, dynamic_column_name="json_column", column_id="foo", value="bar"
    )
    DBSession.refresh(child)
    assert child.aliased_json_column["foo"] == "bar"
    assert child.json_column["foo"] == "bar"


def test_can_update_with_synonym_column():
    parent = _create_obj(Parent)
    atomic_jsonb_set(
        parent, dynamic_column_name="aliased_json_column", column_id="foo", value="bar"
    )
    assert parent.aliased_json_column["foo"] == "bar"
    assert parent.json_column["foo"] == "bar"


def test_can_update_with_synonym_column_on_child():
    child = _create_obj(Child)
    atomic_jsonb_set(
        child, dynamic_column_name="aliased_json_column", column_id="foo", value="bar"
    )
    assert child.aliased_json_column["foo"] == "bar"
    assert child.json_column["foo"] == "bar"

```

**Comment by zzzeek** [comment]
Hi - can you please distill this into a single file [MCVE](http://stackoverflow.com/help/mcve) with no depedenencies  (no flask, no pytest, etc)  and apply inside of a markdown comment?    thanks

**Reply by Amwam** [reply]
@zzzeek Sure thing, I've updated the original description with all the code in a single file (and removed the pytest dependency)

**Reply by zzzeek** [reply]
can you provide the soruce for _create_obj()?  Also hit some syntax errors.

**Reply by Amwam** [reply]
Apologies, its
```python
def _create_obj(cls):
    model = cls()
    DBSession.add(model)
    DBSession.flush()
    return model
```

I'm not seeing any syntax errors, (except for `create_obj()`), where are they appearing?

**Comment by zzzeek** [comment, marked as answer]
> 
> This does make sense as we're passing the column to `DBSession.query()`. 

this should be a mapped class, not a column.  

> Changing this to `context.__class__`/`mapper` (i.e. the table) the tests fail when trying to update the Child class. The emitted SQL targets the child table directly, despite the column being located on the Parent.

I'm not following how you would emit an UPDATE that changes a column on the Parent table, if the target table given is Child.  To my knowledge, PostgreSQL doesn't support "multi-table updates" like MySQL does.    once I get a runnable program I can get some idea what it is you are trying to do.

**Reply by zzzeek** [reply]
that is, you have to provide the mapper or class that is home to the column in question. looking at your mappings, you need to send Parent to update() in all cases.   These caveats are noted at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#update-and-delete-with-arbitrary-where-clause under "Additional Caveats for ORM-enabled updates and deletes":

> ORM-enabled UPDATEs and DELETEs do not handle joined table inheritance automatically. If the operation is against multiple tables, typically individual UPDATE / DELETE statements against the individual tables should be used. 

that note is not actually very clearly written, as you can still use any ORM mapped class that is mapped to a single table, such as the base class in an inheritance hierarchy.

**Reply by Amwam** [reply]
Thanks for your help, thats made things a lot clearer. 
I've managed to get to a solution that now passes one both 1.3 and 1.4, and also works for the synonym properties

```python
def atomic_jsonb_set(
    context, dynamic_column_name: str, column_id, value,
):
    mapper = inspect(context.__class__)
    dynamic_column = getattr(context.__class__, dynamic_column_name)

    query = DBSession.query(mapper.base_mapper.class_)

    query = query.filter(
        *[
            pk_col == pk_value
            for pk_col, pk_value in zip(
                mapper.primary_key, mapper.primary_key_from_instance(context)
            )
        ]
    )

    query.update(
        {
            dynamic_column: func.jsonb_set(
                dynamic_column, f"{{{column_id}}}", json.dumps(value),
            )
        },
        synchronize_session="fetch",
    )
```

**Reply by zzzeek** [reply]
OK cool.  im surprised 1.3 let you send a column to the query, I guess it was diving into it and finding the parent table.  Been trying to eliminate implicit guesswork like that.
